### PR TITLE
Move JsonConfigurationJob to Jobs.Common / Migrate ArchivePackages to JsonConfig

### PR DESCRIPTION
**Pull request by chenriksson**
Part of https://github.com/NuGet/Engineering/issues/1560. Depends on https://github.com/NuGet/NuGet.Jobs/pull/505

Includes the following:
- Moves bulk of `JsonConfigurationJob` into `NuGet.Jobs.Common`, for use by legacy jobs
- Updates validation jobs use db helpers from https://github.com/NuGet/NuGet.Jobs/pull/505
  - `JsonConfigurationJob` registers databases when `GalleryDbConfiguration` or `ValidationDbConfiguration` is found
- Updates `ArchivePackages.Job` to descend from `JsonConfigurationJob`
  - Requires new "Configuration" and "GalleryDatabaseAddress" Octopus config. Will remove old config only after these changes RI'd to master and deployed to prod.

Testing should include the following jobs:
- ArchivePackages (Verified on DEV)
- Revalidate
- ProcessSignature
- RebootSearchInstance
- SubscriptionProcessor
- RevalidateCertificate

### CODE REVIEW

File: src/ArchivePackages/ArchivePackages.Job.cs [hunk 1]
```diff
@@ -66,33 +74,34 @@ public class Job : JobBase
...
         public override void Init(IServiceContainer serviceContainer, IDictionary<string, string> jobArgsDictionary)
         {
-            var secretInjector = (ISecretInjector)serviceContainer.GetService(typeof(ISecretInjector));
-            var packageDbConnectionString = JobConfigurationManager.GetArgument(jobArgsDictionary, JobArgumentNames.PackageDatabase);
-            _packageDbConnectionFactory = new AzureSqlConnectionFactory(packageDbConnectionString, secretInjector);
+            base.Init(serviceContainer, jobArgsDictionary);
 
-            Source = CloudStorageAccount.Parse(
-                        JobConfigurationManager.GetArgument(jobArgsDictionary, JobArgumentNames.Source));
+            Configuration = _serviceProvider.GetRequiredService<InitializationConfiguration>();
```
Comment (chenriksson) [inline src/ArchivePackages/ArchivePackages.Job.cs]: All command-line options for legacy jobs will be moved to a custom `InitializationConfiguration`, from json config.

File: src/ArchivePackages/ArchivePackages.Job.cs [hunk 2]
```diff
@@ -193,129 +202,14 @@ private async Task ArchivePackage(string sourceBlobName, string destinationBlobN
...
+        protected override void ConfigureJobServices(IServiceCollection services, IConfigurationRoot configurationRoot)
         {
-            Published = published;
+            ConfigureInitializationSection<InitializationConfiguration>(services, configurationRoot);
         }
-        public string Id { get; set; }
-        public string Version { get; set; }
-        public string Hash { get; set; }
-        public DateTime? LastEdited { get; set; }
-        public DateTime? Published { get; set; }
```
Comment (chenriksson) [inline src/ArchivePackages/ArchivePackages.Job.cs]: Cleanup - just moving classes into separate source files.

File: src/ArchivePackages/ArchivePackages.csproj [hunk 3]
```diff
@@ -75,16 +82,14 @@
     <PackageReference Include="Newtonsoft.Json">
       <Version>9.0.1</Version>
     </PackageReference>
-    <PackageReference Include="NuGet.Services.Sql">
-      <Version>2.27.0</Version>
-    </PackageReference>
```
Comment (chenriksson) [inline src/ArchivePackages/ArchivePackages.csproj]: We only need direct dependency from `Jobs.Common` now.

File: src/ArchivePackages/Scripts/ArchivePackages.cmd [hunk 4]
```diff
@@ -7,7 +7,7 @@ cd bin
 
 	title #{Jobs.archivepackages.Title}
 
-    start /w archivepackages.exe -VaultName "#{Deployment.Azure.KeyVault.VaultName}" -ClientId "#{Deployment.Azure.KeyVault.ClientId}" -CertificateThumbprint "#{Deployment.Azure.KeyVault.CertificateThumbprint}" -PackageDatabase "#{Jobs.archivepackages.PackageDatabase}" -Source "#{Jobs.archivepackages.Source}" -PrimaryDestination "#{Jobs.archivepackages.PrimaryDestination}" -SecondaryDestination "#{Jobs.archivepackages.SecondaryDestination}" -Sleep "#{Jobs.archivepackages.Sleep}" -InstrumentationKey "#{Jobs.archivepackages.ApplicationInsightsInstrumentationKey}"
+	start /w archivepackages.exe -Configuration "#{Jobs.archivepackages.Configuration}"
```
Comment (chenriksson) [inline src/ArchivePackages/Scripts/ArchivePackages.cmd]: Will add "-Configuration" octopus config now, and remove old config after RI and prod deployments are complete.

File: src/NuGet.Jobs.Common/JsonConfigurationJob.cs [hunk 5]
```diff
@@ -21,26 +17,26 @@
...
-using NuGet.Services.ServiceBus;
-using NuGet.Services.Sql;
-using NuGet.Services.Validation;
-using NuGetGallery;
-using NuGetGallery.Diagnostics;
 
-namespace NuGet.Jobs.Validation
+namespace NuGet.Jobs
 {
     public abstract class JsonConfigurationJob : JobBase
```
Comment (chenriksson) [inline src/NuGet.Jobs.Common/JsonConfigurationJob.cs]: Old `JsonConfigurationJob` is now split into Jobs.Common `JsonConfigurationJob` and Validation.Common.Job `ValidationJobBase`, based on needs of legacy / validation jobs.

File: src/ArchivePackages/Settings/dev.json [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    "ConnectionString": "Data Source=tcp:#{Jobs.validation.GalleryDatabaseAddress};Initial Catalog=nuget-dev-0-v2gallery;Integrated Security=False;User ID=$$Dev-GalleryDBReadOnly-UserName$$;Password=$$Dev-GalleryDBReadOnly-Password$$;Connect Timeout=30;Encrypt=True"
+  },
+
+  "KeyVault_VaultName": "#{Deployment.Azure.KeyVault.VaultName}",
+  "KeyVault_ClientId": "#{Deployment.Azure.KeyVault.ClientId}",
+  "KeyVault_CertificateThumbprint": "#{Deployment.Azure.KeyVault.CertificateThumbprint}",
+  "KeyVault_ValidateCertificate": true,
+  "KeyVault_StoreName": "My",
+  "KeyVault_StoreLocation": "LocalMachine"
+}
```
Comment (chenriksson) [inline src/ArchivePackages/Settings/dev.json]: Job config could eventually move to `NuGetDeployment`, as we do with Orchestrator. For now, I don't want to modify build / deployment infrastructure.

File: src/ArchivePackages/Scripts/ArchivePackages.cmd (hunk 4 above)
Comment (joelverhagen) [inline src/ArchivePackages/Scripts/ArchivePackages.cmd]: Oh nice, so instrumentation key now comes through JSON configuration? I think some of the validation jobs need this fix right?

File: src/ArchivePackages/Settings/dev.json [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+  "Initialization": {
+    "Source": "DefaultEndpointsProtocol=https;AccountName=nugetdevlegacy;AccountKey=$$Dev-NuGetDevLegacyStorage-Key$$",
+    "PrimaryDestination": "DefaultEndpointsProtocol=https;AccountName=nugetdevlegacy;AccountKey=$$Dev-NuGetDevLegacyStorage-Key$$",
+    "SecondaryDestination": "DefaultEndpointsProtocol=https;AccountName=nugetdev1;AccountKey=$$Dev-NuGetDev1Storage-Key$$",
+    "Sleep": "600000",
+    "InstrumentationKey": "$$Dev-ApplicationInsightsV2Gallery-InstrumentationKey$$"
+  },
+
+  "GalleryDb": {
+    "ConnectionString": "Data Source=tcp:#{Jobs.validation.GalleryDatabaseAddress};Initial Catalog=nuget-dev-0-v2gallery;Integrated Security=False;User ID=$$Dev-GalleryDBReadOnly-UserName$$;Password=$$Dev-GalleryDBReadOnly-Password$$;Connect Timeout=30;Encrypt=True"
```
Comment (chenriksson) [inline src/ArchivePackages/Settings/dev.json]: Will fix `GalleryDatabaseAddress` prefix

File: src/Validation.Common.Job/ValidationJobBase.cs [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+            ServicePointManager.DefaultConnectionLimit = MaximumConnectionsPerServer;
+        }
+
+        protected override void ConfigureDefaultJobServices(IServiceCollection services, IConfigurationRoot configurationRoot)
+        {
+            base.ConfigureDefaultJobServices(services, configurationRoot);
+
+            //services.AddSingleton(new TelemetryClient());
+            //services.AddTransient<ITelemetryClient, TelemetryClientWrapper>();
```
Comment (chenriksson) [inline src/Validation.Common.Job/ValidationJobBase.cs]: Will remove and keep in JsonConfigurationJob

File: src/ArchivePackages/Scripts/ArchivePackages.cmd (hunk 4 above)
Comment (chenriksson) [reply src/ArchivePackages/Scripts/ArchivePackages.cmd]: The `Revalidate` job now has an "AppInsights" section. I should probably extract that into `Jobs.Common and just reuse it. I had just planned to copy all command-line arguments to a job-specific `InitializationConfiguration`.

**Pull request merged**